ompi/mca/osc/ucx/osc_ucx_component.c

```diff
@@ -0,0 +1,565 @@
+        ompi_osc_ucx_fetch_and_op,
+        ompi_osc_ucx_get_accumulate,
+
+        ompi_osc_ucx_rput,
+        ompi_osc_ucx_rget,
+        ompi_osc_ucx_raccumulate,
+        ompi_osc_ucx_rget_accumulate,
+
+        ompi_osc_ucx_fence,
+
+        ompi_osc_ucx_start,
+        ompi_osc_ucx_complete,
+        ompi_osc_ucx_post,
+        ompi_osc_ucx_wait,
+        ompi_osc_ucx_test,
+
+        ompi_osc_ucx_lock,
+        ompi_osc_ucx_unlock,
+        ompi_osc_ucx_lock_all,
+        ompi_osc_ucx_unlock_all,
+
+        ompi_osc_ucx_sync,
+        ompi_osc_ucx_flush,
+        ompi_osc_ucx_flush_all,
+        ompi_osc_ucx_flush_local,
+        ompi_osc_ucx_flush_local_all,
+
+        ompi_osc_ucx_set_info,
+        ompi_osc_ucx_get_info
+    }
+};
+
+static int component_open(void) {
+    return OMPI_SUCCESS;
+}
+
+static int component_register(void) {
+    return OMPI_SUCCESS;
+}
+
+static int component_init(bool enable_progress_threads, bool enable_mpi_threads) {
+    ucp_config_t *config = NULL;
+    ucp_params_t context_params;
+    int ret = OMPI_SUCCESS;
+    ucs_status_t status;
+
+    mca_osc_ucx_component.ucp_context = NULL;
+    mca_osc_ucx_component.ucp_worker = NULL;
+    mca_osc_ucx_component.enable_mpi_threads = enable_mpi_threads;
+
+    status = ucp_config_read("MPI", NULL, &config);
+    if (UCS_OK != status) {
+        return OMPI_ERROR;
+    }
+
+    /* initialize UCP context */
+
+    memset(&context_params, 0, sizeof(ucp_context_h));
+    context_params.field_mask = UCP_PARAM_FIELD_FEATURES |
+                                UCP_PARAM_FIELD_MT_WORKERS_SHARED |
+                                UCP_PARAM_FIELD_ESTIMATED_NUM_EPS |
+                                UCP_PARAM_FIELD_REQUEST_INIT |
+                                UCP_PARAM_FIELD_REQUEST_SIZE;
+    context_params.features = UCP_FEATURE_RMA | UCP_FEATURE_AMO32 | UCP_FEATURE_AMO64;
+    context_params.mt_workers_shared = 0;
+    context_params.estimated_num_eps = ompi_proc_world_size();
+    context_params.request_size = sizeof(ompi_osc_ucx_request_t);
+    context_params.request_init = request_init;
+
+    status = ucp_init(&context_params, config, &mca_osc_ucx_component.ucp_context);
+    ucp_config_release(config);
+    if (UCS_OK != status) {
+        return OMPI_ERROR;
+    }
+
+    return ret;
+}
+
+static int component_finalize(void) {
+    int i;
+    for (i = 0; i < ompi_proc_world_size(); i++) {
+        ucp_ep_h ep = OSC_UCX_GET_EP(&(ompi_mpi_comm_world.comm), i);
+        if (ep != NULL) {
+            ucp_ep_destroy(ep);
+        }
+    }
+
+    if (mca_osc_ucx_component.ucp_worker != NULL) {
+        ucp_worker_destroy(mca_osc_ucx_component.ucp_worker);
+    }
+
+    ucp_cleanup(mca_osc_ucx_component.ucp_context);
+    return OMPI_SUCCESS;
+}
+
+static int component_query(struct ompi_win_t *win, void **base, size_t size, int disp_unit,
+                           struct ompi_communicator_t *comm, struct ompi_info_t *info, int flavor) {
+    return OMPI_SUCCESS;
```

Comment:
(now L203)
I temporarily set priority as 100 (see rdma's prioi is 90). Is that OK or we should have other priority number?